### PR TITLE
Allow OpenMPConstruct evaluations in PFT to be targets of branch

### PR DESCRIPTION
**Pull request by kiranchandramohan**
This patch allows Directives (OpenMPConstruct, OpenACCConstruct etc) to
be target of branch instructions. If an OpenMPConstruct follows an
unstructured construct then the OpenMPConstruct can be the target of the
branch that exits from the unstructured construct. This was not
previously possible.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
Was not sure what the right way to fix this issue. Please let me know whether I should approach this in any other way.

Previously posted in Slack Channel: (@schweitzpgi pointed out that this is true for all constructs, but constructs usually have begin and end construct statements that cover this.)
Restrictions on Directives (like OpenMPConstruct) in the PFT is causing a problem when the code that is immediately following an unstructured loop is an OpenMP construct, like in the example below.
```
PROGRAM im
  LOGICAL :: inrdone
  INTEGER :: i
  do i = 1, 100
  goto 2
2 inrdone = .false.
  end do

  !$OMP MASTER
  inrdone = .TRUE.
  !$OMP END MASTER
END PROGRAM im
```
 The FIR generated for this code will have the code inside the master directive outside the `omp.master` region. I think this happens (as can be seen from the PFT below) because `omp.master `is not the target of the branch but `inrdone = .TRUE` is.
```
1 Program im
  <<DoConstruct!>> -> 5
    1 NonLabelDoStmt -> 4: do i = 1, 100
    2 ^GotoStmt! -> 3: goto 2
    3 ^AssignmentStmt: 2 inrdone = .false.
    4 EndDoStmt -> 1: end do
  <<End DoConstruct!>>
  <<OpenMPConstruct>>
    5 ^AssignmentStmt: inrdone = .true.
  <<End OpenMPConstruct>>
  6 EndProgramStmt: end program im
End Program im
```

**Review by schweitzpgi — COMMENTED** [review]
Seems simple and reasonable to me. I don't know if @vdonaldson had a deeper reason behind the original implementation.

File: flang/lib/Lower/PFTBuilder.cpp [hunk 1]
```diff
@@ -996,7 +996,10 @@ class PFTDumper {
     llvm::StringRef name = evaluationName(eval);
     std::string bang = eval.isUnstructured ? "!" : "";
     if (eval.isConstruct() || eval.isDirective()) {
-      outputStream << indentString << "<<" << name << bang << ">>";
+      outputStream << indentString ;
```
Comment (clementval) [inline flang/lib/Lower/PFTBuilder.cpp:999]:
```suggestion
      outputStream << indentString;
```

Comment (kiranchandramohan) [reply flang/lib/Lower/PFTBuilder.cpp]: Thanks. Done.

**Review by clementval — APPROVED** [review]
LGTM

**Comment by schweitzpgi** [comment]
@kiranchandramohan 

@vdonaldson is looking into this problem at the moment. Stay tuned...

**Comment by kiranchandramohan** [comment]
Ping. @vdonaldson @schweitzpgi Do you have any high-level comments here?  Like whether this is the right way to go or do we need some larger changes to make sure that lowering works correctly with OpenMP/OpenACC?

We might also have a similar issue in https://github.com/flang-compiler/f18-llvm-project/issues/1120.

**Review by vdonaldson — APPROVED** [review]

**Comment by kiranchandramohan** [comment]
Thanks @vdonaldson for the review and approval. @schweitzpgi can this be merged?

> So goto, cycle, exit, arithmetic goto, I/O calls with ERR, EOR, EOF and iostat specifiers, calls with alternate returns, and a few other statements that operate local to an OpenMP construct are valid. All of these are "unstructured" in FIR, and must be implemented directly with branches between generic basic blocks. The structured fir.do_loop and fir.if ops cannot contain any of these branches.

For the cycle case can we have a fir.if guarding the code after the cycle? This way the loop can be represented by a fir.do_loop and the omp.ws_loop can reuse the code for the lowering. And we can solve the particular case of https://github.com/flang-compiler/f18-llvm-project/issues/1120.

**Comment by vdonaldson** [comment]
> For the cycle case can we have a fir.if guarding the code after the cycle? This way the loop can be represented by a fir.do_loop and the omp.ws_loop can reuse the code for the lowering. And we can solve the particular case of #1120.

A `cycle` statement may appear at any level of nesting inside a `do` loop, including inside a nested `do`.  It isn't in general possible to replace a `cycle` statement with a single `if` statement.

**Comment by kiranchandramohan** [comment]
Also, we seem to generate the fir.do_loop with fir.if for the following code as noted by @Leporacanthicus . Isn't the cycle case in the previous comment similar?

```
program n
  integer :: i
  do i = 1,5
     if (i == 1) goto 20
     print *,i
 20 end do
end program n
```

```
  %3 = fir.do_loop %arg0 = %1 to %2 step %c1 -> index {
    %5 = fir.convert %arg0 : (index) -> i32
    fir.store %5 to %0 : !fir.ref<i32>
    %6 = fir.load %0 : !fir.ref<i32>
    %c1_i32_0 = constant 1 : i32
    %7 = cmpi eq, %6, %c1_i32_0 : i32
    %true = constant true
    %8 = xor %7, %true : i1
    fir.if %8 {
      %c-1_i32 = constant -1 : i32
      %10 = fir.address_of(@_QQcl.2E2F737472756374757265642E66393000) : !fir.ref<!fir.char<1,17>>
      %11 = fir.convert %10 : (!fir.ref<!fir.char<1,17>>) -> !fir.ref<i8>
      %c5_i32_1 = constant 5 : i32
      %12 = fir.call @_FortranAioBeginExternalListOutput(%c-1_i32, %11, %c5_i32_1) : (i32, !fir.ref<i8>, i32) -> !fir.ref<i8>
      %13 = fir.load %0 : !fir.ref<i32>
      %14 = fir.convert %13 : (i32) -> i64
      %15 = fir.call @_FortranAioOutputInteger64(%12, %14) : (!fir.ref<i8>, i64) -> i1
      %16 = fir.call @_FortranAioEndIoStatement(%12) : (!fir.ref<i8>) -> i32
    } else {
    }
    %9 = addi %arg0, %c1 : index
    fir.result %9 : index
  }
```

**Comment by schweitzpgi** [comment]
@kiranchandramohan @vdonaldson 
We don't really have a strong case at the moment to go the heroic route of recovering structured control flow from any general control-flow graph. On the other hand, I'm not sure that's really what's needed.

Is there any way to classify a subset for what's needed for OpenMP? Maybe constrained to the "interior" within the innermost loop of a perfect loop nest or something like that? (More precisely, edges that skip to the loop header or skip some subset of statements in the loop body.)

**Comment by kiranchandramohan** [comment]
@schweitzpgi I think from what we have seen in the SNAP application, we are interested in cases where there is a `cycle` statement in the `loop` inside an `if `statement.
-> `Cycle` statement that applies to the innermost `loop` which is inside an `if` statement in the top level of the `loop`

As @Leporacanthicus points out in https://github.com/flang-compiler/f18-llvm-project/issues/1120, `Exit` statements in similar situations are also interesting for us. Can something be done for this as well?
```
  do i = 1,5
     if (i == 4) exit
     print *,i
  end do
```

**Pull request closed without merging**